### PR TITLE
Feature/fortran api redistribution

### PR DESCRIPTION
**Pull request by sbrdar**
Add Fortran API for atlas::Redistribution
(with Willem D.)

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/atlas?pullRequest=160) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/atlas?pullRequest=160) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/atlas?pullRequest=160) it.</sub>

File: src/atlas_f/redistribution/atlas_Redistribution_module.F90 [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+end function
+
+function ctor_create( fspace1, fspace2, redist_name ) result(this)
+  use atlas_redistribution_c_binding
+  class(atlas_FunctionSpace), intent(in) :: fspace1, fspace2
+  character(len=*), intent(in), optional :: redist_name
+  type(atlas_Redistribution) :: this
+  type(atlas_Config) :: config
+  config = empty_config()
+  call config%set("type", "RedistributeGeneric")
```
Comment (wdeconinck) [inline src/atlas_f/redistribution/atlas_Redistribution_module.F90:87]: I would remove this line `call config%set("type", "RedistributeGeneric")`
and allow to pass down an empty config.

File: src/atlas/redistribution/detail/RedistributionInterface.cc [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+// Fortran interfaces
+// ----------------------------------------------------------------------------
+
+extern "C" {
+
+detail::RedistributionImpl* atlas__Redistribution__new__config(
+    const functionspace::FunctionSpaceImpl* fspace1, const functionspace::FunctionSpaceImpl* fspace2, 
+    const eckit::Configuration* config) {
+    ATLAS_ASSERT(config != nullptr);
+    std::string type;
```
Comment (wdeconinck) [inline src/atlas/redistribution/detail/RedistributionInterface.cc:31]:
```suggestion
    std::string type = detail::RedistributeGeneric::static_type();
```
and add include `#include "atlas/redistribution/detail/RedistributeGeneric.h"` below line 16

**Review by wdeconinck — CHANGES_REQUESTED** [review]
Very good! I have some minor suggestions only.
Further to in-code suggestions, please in atlas_module, add a line to export atlas_redistribution type.

File: src/atlas/redistribution/detail/RedistributionInterface.cc (hunk 2 above)
Comment (wdeconinck) [reply src/atlas/redistribution/detail/RedistributionInterface.cc:31]: This catches passing down the empty config in Fortran

**Comment by wdeconinck** [comment]
Please also address failed compilation with Intel compiler, see e.g. https://github.com/ecmwf/atlas/actions/runs/6708816197/job/18230465722?pr=160#step:14:2303

File: src/tests/redistribution/CMakeLists.txt [hunk 3]
```diff
@@ -7,6 +7,15 @@
 
 if( atlas_HAVE_ATLAS_FUNCTIONSPACE )
```
Comment (wdeconinck) [inline src/tests/redistribution/CMakeLists.txt:9]: The `add_fctest` command needs to be guarded by 
```
if( HAVE_FORTRAN )
  add_fctest(...)
endif()

File: src/atlas_f/redistribution/atlas_Redistribution_module.F90 (hunk 1 above)
Comment (sbrdar) [reply src/atlas_f/redistribution/atlas_Redistribution_module.F90:87]: This config%set has been removed now. I had to create a default case to be "RedistributeGeneric".

File: src/atlas/redistribution/detail/RedistributionInterface.cc (hunk 2 above)
Comment (sbrdar) [reply src/atlas/redistribution/detail/RedistributionInterface.cc:31]: Done.

**Review by wdeconinck — APPROVED** [review]
All good now. Thanks!

**Pull request merged**